### PR TITLE
Auth endpoints

### PR DESCRIPTION
**Pull request by silverware13**
Back-end:
Integrate authentication usage for the plan, comment, course, and review API endpoints.
Add additional constraints.
Refactor createPlan and updatePlan functions.
Remove some unused API endpoints.

Front-end:
Add ability to delete plans (with warning).
Improve page error handling.
Have "create review" selector disable unusable options.
Have delete plan, edit plan, create review, and create comment conditionally render based on if they are currently usable.

### CODE REVIEW

File: api/comment.js [hunk 1]
```diff
@@ -61,28 +60,4 @@ app.post("/", requireAuth, async (req, res) => {
...
-      res.status(200).send(results);
-    }
-
-  } catch (err) {
-    console.error("500: An internal server error occurred\n Error:", err);
-    res.status(500).send({error: "An internal server error occurred. Please try again later."});
-  }
-
-});
-
```
Comment (philectron) [inline api/comment.js:87]: Why did you remove this? I guess that it's unused?

File: api/comment.js [hunk 2]
```diff
@@ -6,7 +6,6 @@ const express = require("express");
 const app = express();
 const {requireAuth} = require("../services/auth/auth");
 const enforceConstraints = require("../services/validation/commentValidation").enforceConstraints;
-const getComment = require("../models/comment").getComment;
 const createComment = require("../models/comment").createComment;
```
Comment (philectron) [inline api/comment.js]: If you only want to import a single function (or multiple functions) from a module, it looks better if you do it like this:

```js
const {function1, function2, ...} = require("./some/module");
```

File: api/plan.js [hunk 3]
```diff
@@ -4,49 +4,57 @@
...
-const enforceConstraints = require("../services/validation/planValidation").enforceConstraints;
-const patchEnforceConstraints = require("../services/validation/planValidation").patchEnforceConstraints;
-const savePlan = require("../models/plan").savePlan;
-const updatePlan = require("../models/plan").updatePlan;
-const getPlan = require("../models/plan").getPlan;
-const getPlansStatus = require("../models/plan").getPlansStatus;
-const deletePlan = require("../models/plan").deletePlan;
-const getPlanActivity = require("../models/plan").getPlanActivity;
 const {requireAuth} = require("../services/auth/auth");
+const formatStringArray = require("../services/format/format").formatStringArray;
```
Comment (philectron) [inline api/plan.js]:
```js
const {formatStringArray} = require("../services/format/format");
```

File: models/comment.js [hunk 4]
```diff
@@ -3,23 +3,6 @@
...
-    return results[0];
-
-  } catch (err) {
-    console.log("Error searching for comment");
-    throw Error(err);
-  }
-
-}
-exports.getComment = getComment;
-
```
Comment (philectron) [inline models/comment.js:22]: Unused?

File: models/plan.js [hunk 5]
```diff
@@ -4,86 +4,116 @@
...
+async function createPlan(userId, planName, courses) {
 
   try {
 
-    // save the basic plan information
-    const planId = await insertPlan(userId, planName);
+    // construct the first SQL query
+    let sql = "BEGIN; " +
+      "INSERT INTO Plan (studentId, planName, status) VALUES (?, ?, 2); " +
+      "SELECT LAST_INSERT_ID();";
```
Comment (philectron) [inline models/plan.js:14]: I wonder why you did this, since the result of the `INSERT INTO` query is the `insertId` of the new object.

**Review by philectron — COMMENTED** [review]
Looks good. I skimmed through the React and CSS stuff.

File: api/comment.js (hunk 1 above)
Comment (silverware13) [reply api/comment.js:87]: Yes

**Comment by philectron** [comment]
Also, I'm guessing from this PR that the User authorization/permission feature is left for another PR?

File: models/comment.js (hunk 4 above)
Comment (silverware13) [reply models/comment.js:22]: yep

File: models/plan.js (hunk 5 above)
Comment (silverware13) [reply models/plan.js:14]: It is a transaction. While it looks like a single insert query, it is not. I will not get the insert value back until EVERYTHING is ran, meaning I need someway to store the ID mid transaction. This is the way to do that. This allows me to insert the base plan and the selected courses atomically, meaning that if the database crashes halfway through the process we don't end-up with a plan with no courses.

**Review by philectron — APPROVED** [review]
